### ISSUE TITLE
Ensure history integrity for audit use case

### ISSUE THREAD

**Issue by mscansian** [treyhunner/django-simple-history#420]
### Use case
We are a small company that develops a web app for the financial market. We have several Django models that we need to track the changes for audit purposes. We are required by law to do that and sometimes a customer needs to know who/when/how changed the parameters of a specific contract, setting, etc. Currently *django-simple-history* is fulfilling our needs. It keeps a version of all changes on the selected models in the database and we can check at any point in time.

### Problem
In order to have a true audit system we need to ensure that the model history cannot be changed or deleted. Since *django-simple-history* uses the application database as storage, it's quite easy to update or delete a record. We need something that restrict this kind of action to a tiny subset of people. Also, since these logs are part of our DR plan, we need to have an immutable off-site backup.

### What I have considered
Overriding `create_historical_record` to send the information to a third-party service (such as DynamoDB, CloudWatchLogs or Kinesis). This means that we can have the best of both worlds: The nice history view that *django-simple-history* provides and also a secure and immutable backup for audit and DR.

At first we though about forking the project and implementing this for internal use. I'm not sure, but I think this kind of functionality is overextending the scope of the project, right? I don't wanna do that. Forking means duplicating the effort to maintain a codebase and provides no benefit to the upstream project.

### What I need
I need to understand if there's a way to implement this that we can merge in the main project. I'm not sure what are the boundaries of the project's scope. If you guys think that this feature is a nice addition, I will gladly send a PR :)

**Comment by rossmechanic** [comment]
Hey @mscansian. Great to hear about your company using DSH as an audit-log. It's super helpful for that purpose. I think that your use case is a good one, and is definitely within the scope of DSH. However, I think we should discuss the best way of implementing it. At first I thought you might just be able to use a `post_save` signal on the historical model, but you'd have to register one for each of your historical models, and that doesn't exactly seem DRY. What I'm thinking here is that we add a hook in the `create_historical_record` function that allows a developer to write custom functionality that occurs when an historical record is being created. 

Basically, I'm happy to accept a PR that's generalizable and backward-compatible which allows you to plugin your own process for sending the information to a third-party. What do you think?

**Comment by mscansian** [comment]
Hi @rossmechanic. That's exactly the kind of discussion I intended to have here :) Having a hook in the `create_historical_record` would meet our needs. As you said, I could implement the code to push the changes to another service within our application. This has the added benefit that in the future we can change the way it works as we please.

I'm thinking about creating two custom [Signals](https://docs.djangoproject.com/en/2.1/topics/signals/) and sending it in `create_historical_record`. Probably `pre_create_historical_record` and `post_create_historical_record`. We only need one, but it could benefit other use cases. Do you see any reason to implement it differently?

**Comment by rossmechanic** [comment]
`pre_create_historical_record` and `post_create_historical_record` sound good to me. Feel free to put up a PR. Thanks!

**Comment by rossmechanic** [comment]
Closed by #426
